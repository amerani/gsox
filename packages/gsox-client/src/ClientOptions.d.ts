declare class ClientOptions {
      public host: string;
      public port: number;
      public routes: {
            graphql: string
            webhook: string,
      };
      public ws: any;
}
