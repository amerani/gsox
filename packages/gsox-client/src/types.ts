import { Field, Type } from "@gsox/schema";

@Type()
class Notification {
      @Field("Int")
      public id: number;
      @Field()
      public type: string;
      @Field()
      public timestamp: string;
}

@Type()
class Stream {
      @Field("Int")
      public id: number;
      @Field()
      public type: string;
      @Field()
      public data: string;
}

@Type()
class AlertType {
      @Field()
      public type: string;
}

@Type()
class Alert {
      @Field("Int")
      public id: number;
      @Field()
      public message: string;
      @Field()
      public severity: string;
      @Field()
      public timestamp: string;
      @Field(AlertType)
      public type: AlertType;
}

@Type()
class Complex {
      @Field("Int")
      public id;
      @Field(Alert)
      public alert;
}

export {
      Notification, Alert, Stream, Complex,
};
