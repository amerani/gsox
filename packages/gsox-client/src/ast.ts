import {
      DocumentNode,
      FieldNode,
      NameNode,
      OperationDefinitionNode,
      SelectionNode,
      SelectionSetNode,
    } from "graphql";

class SubscriptionDocument implements DocumentNode {
      public kind: "Document" = "Document";
      public definitions: OperationDefinitionNode[];
      constructor(definitions) {
            this.definitions = definitions;
            this.kind = "Document";
      }
}

class OperationNode implements OperationDefinitionNode {
      public kind: "OperationDefinition";
      public operation: "subscription";
      public selectionSet: SelectionSetNode;
      constructor(selectionSet) {
            this.selectionSet = selectionSet;
            this.kind = "OperationDefinition";
            this.operation = "subscription";
      }
}

class SelectionSet implements SelectionSetNode {
      public kind: "SelectionSet";
      public selections: FieldNode[];
      constructor(selections: FieldNode[]) {
            this.selections = selections;
            this.kind = "SelectionSet";
      }
}

class NameNodeImp implements NameNode {
      public kind: "Name";
      public value: string;
      constructor(value) {
            this.value = value;
            this.kind = "Name";
      }
}

class Field implements FieldNode {
      public kind: "Field";
      public name: NameNode;
      public selectionSet?: SelectionSetNode;
      constructor(name, selectionSet?) {
            this.name = new NameNodeImp(name);
            this.selectionSet = selectionSet;
            this.kind = "Field";
      }
}

export { Field, SelectionSet, OperationNode, SubscriptionDocument };
