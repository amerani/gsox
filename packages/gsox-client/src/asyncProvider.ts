import { InMemoryCache } from "apollo-cache-inmemory";
import { ApolloClient} from "apollo-client";
import { WebSocketLink } from "apollo-link-ws";
import { gql } from "apollo-server-express";
import { SubscriptionClient } from "subscriptions-transport-ws";
import * as ws from "ws";

const port = 3000;
const wsClient = new SubscriptionClient(`ws://localhost:${port}/graphql`, { reconnect: true }, ws);
const client = new ApolloClient({
      cache: new InMemoryCache({}),
      link: new WebSocketLink(wsClient),
});
client.subscribe({
      query: gql`
      subscription {
            msg {
                  type
                  id

            }
            }`,
}).subscribe(
      console.log,
      console.log,
      console.log,
);
