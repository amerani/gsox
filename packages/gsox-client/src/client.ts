import { InMemoryCache } from "apollo-cache-inmemory";
import ApolloClient from "apollo-client";
import { ApolloLink } from "apollo-link";
import { WebSocketLink } from "apollo-link-ws";
import { SubscriptionClient } from "subscriptions-transport-ws";
const GRAPHQL_ENDPOINT = "ws://localhost:3000/graphql";

const wsClient = new SubscriptionClient(GRAPHQL_ENDPOINT, {
  reconnect: true,
});

const wsLink = new WebSocketLink(wsClient);

export const client = new ApolloClient({
  cache: new InMemoryCache({}),
  connectToDevTools: true,
  link: ApolloLink.from([wsLink]),
  ssrForceFetchDelay: 100,
});
