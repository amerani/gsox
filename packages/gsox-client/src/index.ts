export * from "./createClient";
export * from "./context";
export * from "./typeGraph";
