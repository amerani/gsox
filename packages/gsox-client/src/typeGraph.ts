import { FIELD_SYMBOL, TYPE_SYMBOL } from "@gsox/core";
import "reflect-metadata";

function build(types) {
      const map: any = {};
      const typeDefs = [...Object.values(types)];
      typeDefs.forEach((type: any) => {
            const inst = new type();
            const typeName = Reflect.get(inst, TYPE_SYMBOL);
            const fields = Reflect.getMetadata(FIELD_SYMBOL, inst);
            map[typeName] = fields;
      });
      return map;
}

export { build };
