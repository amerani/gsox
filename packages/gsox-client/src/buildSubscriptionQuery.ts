import { Field, OperationNode, SelectionSet, SubscriptionDocument } from "./ast";

function toGql(fields) {
    const gFields = [];
    for (let i = 0; i < fields.length; i++) {
        const field = new Field(fields[i].name);
        if (fields[i].children.length > 0) {
            field.selectionSet = new SelectionSet(toGql(fields[i].children));
        }
        gFields.push(field);
    }
    return gFields;
}

function build(typeName, fields) {
    const gqlFields = toGql(fields);
    const gqlSub = new Field(typeName, new SelectionSet(gqlFields));
    const op = new OperationNode(new SelectionSet([gqlSub]));
    const doc = new SubscriptionDocument([op]);
    return doc;
}

export { build };
