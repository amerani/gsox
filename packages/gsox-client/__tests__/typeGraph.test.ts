import { build } from "../src/typeGraph";
import * as types from "./types";

let typeGraph;
beforeEach(() => {
      typeGraph = build([...Object.values(types)]);
});

test('should build typegraph', () => {
      expect(typeGraph.Complex).not.toBeNull();
      expect(typeGraph.Alert).not.toBeNull();
      expect(typeGraph).toMatchSnapshot();
});

test('Complex should have Alert and children', () => {
      const complex = typeGraph.Complex;
      expect(complex).not.toBeNull();
      const alert = complex.filter(x => x.type === "Alert");
      expect(alert).toHaveLength(1);
      expect(alert[0].children.length).toBeGreaterThanOrEqual(1);
      expect(alert).toMatchSnapshot();
})

test('Alert should have AlertType', () => {
      const complex = typeGraph.Complex;
      const alert = complex.filter(x => x.type === "Alert")[0];
      expect(alert.children.length).toBeGreaterThanOrEqual(1);
      const alertType = alert.children.filter(x => x.type === "AlertType");
      expect(alertType).toHaveLength(1);
      expect(alertType[0].children.length).toBeGreaterThanOrEqual(1);
      expect(alert.children).toMatchSnapshot();
})