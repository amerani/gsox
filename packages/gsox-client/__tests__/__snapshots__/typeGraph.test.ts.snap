// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Alert should have AlertType 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "children": Array [],
        "id": "type",
        "name": "type",
        "type": "String",
      },
    ],
    "id": "type",
    "name": "type",
    "type": "AlertType",
  },
  Object {
    "children": Array [],
    "id": "timestamp",
    "name": "timestamp",
    "type": "String",
  },
  Object {
    "children": Array [],
    "id": "severity",
    "name": "severity",
    "type": "String",
  },
  Object {
    "children": Array [],
    "id": "message",
    "name": "message",
    "type": "String",
  },
  Object {
    "children": Array [],
    "id": "id",
    "name": "id",
    "type": "Int",
  },
]
`;

exports[`Complex should have Alert and children 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "children": Array [],
            "id": "type",
            "name": "type",
            "type": "String",
          },
        ],
        "id": "type",
        "name": "type",
        "type": "AlertType",
      },
      Object {
        "children": Array [],
        "id": "timestamp",
        "name": "timestamp",
        "type": "String",
      },
      Object {
        "children": Array [],
        "id": "severity",
        "name": "severity",
        "type": "String",
      },
      Object {
        "children": Array [],
        "id": "message",
        "name": "message",
        "type": "String",
      },
      Object {
        "children": Array [],
        "id": "id",
        "name": "id",
        "type": "Int",
      },
    ],
    "id": "alert",
    "name": "alert",
    "type": "Alert",
  },
]
`;

exports[`should build typegraph 1`] = `
Object {
  "Alert": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [],
          "id": "type",
          "name": "type",
          "type": "String",
        },
      ],
      "id": "type",
      "name": "type",
      "type": "AlertType",
    },
    Object {
      "children": Array [],
      "id": "timestamp",
      "name": "timestamp",
      "type": "String",
    },
    Object {
      "children": Array [],
      "id": "severity",
      "name": "severity",
      "type": "String",
    },
    Object {
      "children": Array [],
      "id": "message",
      "name": "message",
      "type": "String",
    },
    Object {
      "children": Array [],
      "id": "id",
      "name": "id",
      "type": "Int",
    },
  ],
  "Complex": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "children": Array [],
              "id": "type",
              "name": "type",
              "type": "String",
            },
          ],
          "id": "type",
          "name": "type",
          "type": "AlertType",
        },
        Object {
          "children": Array [],
          "id": "timestamp",
          "name": "timestamp",
          "type": "String",
        },
        Object {
          "children": Array [],
          "id": "severity",
          "name": "severity",
          "type": "String",
        },
        Object {
          "children": Array [],
          "id": "message",
          "name": "message",
          "type": "String",
        },
        Object {
          "children": Array [],
          "id": "id",
          "name": "id",
          "type": "Int",
        },
      ],
      "id": "alert",
      "name": "alert",
      "type": "Alert",
    },
    Object {
      "children": Array [],
      "id": "id",
      "name": "id",
      "type": "Int",
    },
  ],
  "Notification": Array [
    Object {
      "children": Array [],
      "id": "timestamp",
      "name": "timestamp",
      "type": "String",
    },
    Object {
      "children": Array [],
      "id": "type",
      "name": "type",
      "type": "String",
    },
    Object {
      "children": Array [],
      "id": "id",
      "name": "id",
      "type": "Int",
    },
  ],
  "Stream": Array [
    Object {
      "children": Array [],
      "id": "data",
      "name": "data",
      "type": "String",
    },
    Object {
      "children": Array [],
      "id": "type",
      "name": "type",
      "type": "String",
    },
    Object {
      "children": Array [],
      "id": "id",
      "name": "id",
      "type": "Int",
    },
  ],
}
`;
